### ISSUE TITLE
Tuple + Record Intersection Assignment Compatibility Inconsistency

### ISSUE THREAD

**Issue by harrysolovay** [microsoft/TypeScript#40269]
**TypeScript Version:**  4.1.0-dev.20200824

**Search Terms:** tuple, object, intersection, type, assignment, coercion

**Code**

Here is the type that I'd like to use in an assignment:

```ts
type X = {example: "Field"} & number[];
```

The following produces TS error 2322.

```ts
const x: X = {
  ...[1, 2],
  example: "Field",
} as const;
```

If we update this example like so, the type error disappears:

```diff
- const x: X = {
+ const x = {
    ...[1, 2],
    example: "Field",
  } as const;

+ const y: X = x;
```

**Expected behavior:**

I would have expected the first assignment to not produce a type error. I'm wondering if this inconsistency (if that's the correct word for it) is an edge case, or by design? Apologies if this issue is unhelpful! Thank you!

**Actual behavior:**

**Playground Link:** [https://www.typescriptlang.org/play?#code/HYQwtgpgzgDiDGEAEAhATgewNYWEg3gLABQSSALgJ4zIAaSAvARAB7gwA2EAXEgEQAxAJYQOAEz4BfJADIkwAK5gARhDQBtALoBuEiTLwMwKOSQte9JkVJkkAOgfqAjABokAJk0v9t1uy68giLifN420iBQSIbG5LrEknrEoJCwCMgA6hhoWFHWZFQ0SJbMbGCcPPzCohLScooqalrxPjEmZowEPmQOds5unmG2SH7lAVXBEkNIEVFtcUkGRu2UFp0s8ZJAA](https://www.typescriptlang.org/play?#code/HYQwtgpgzgDiDGEAEAhATgewNYWEg3gLABQSSALgJ4zIAaSAvARAB7gwA2EAXEgEQAxAJYQOAEz4BfJADIkwAK5gARhDQBtALoBuEiTLwMwKOSQte9JkVJkkAOgfqAjABokAJk0v9t1uy68giLifN420iBQSIbG5LrEknrEoJCwCMgA6hhoWFHWZFQ0SJbMbGCcPPzCohLScooqalrxPjEmZowEPmQOds5unmG2SH7lAVXBEkNIEVFtcUkGRu2UFp0s8ZJAA)

**Comment by harrysolovay** [comment]
Seems that overriding `sort`, `fill` and `copyWithin` as any blocks the error:

```ts
const x: X = {
  ...[1, 2],
  example: "Field",
  sort: undefined as any,
  fill: undefined as any,
  copyWithin: undefined as any,
} as const;
```

Still curious why the latter example works properly... but this is likely best suited for StackOverflow. Thank you anyways!

**Comment by harrysolovay** [comment]
After some more digging I think it's worth reopening this issue.

Upon omitting the `sort` method from `number[]`, there's no type error:

```ts
type X = {example: "Field"} & Omit<number[], "sort">;

const x: X = {
  ...[0, 1],
  example: "Field",
} as const;
```

Let's now update the code above to re-include `number[]["sort"]`:

```ts
type X = {example: "Field"} & Omit<number[], "sort"> & {sort: number[]["sort"]};

const x: X = {
  ...[0, 1],
  example: "Field",
} as const;
```

^ this produces no type error... so why is it that the initial attempt does?
